llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +152,286 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+    Args.CounterArgSet.push_back(IntrinsicCall->getArgOperand(Op));
+  }
+  return Args;
+}
+
+// We can reach a pointer definition starting from a counter.
+static bool isDirectReachable(SmallVector<const Value *, 5> CounterArgSet,
+                              const Value *V) {
+
+  for (auto Counter : CounterArgSet) {
+    if (V != lookThroughAIEAddressingModes(Counter))
+      return false;
+  }
+
+  return true;
+}
+
+/// This function returns some tracking information related to the upward path
+/// that starts from a Value *TargetPointer to the Phi node that firstly defines
+/// a pointer value to the same ambiguous memory. We start the search from the
+/// last update of the pointer.
+std::optional<IntrinsicChainInfo>
+trackIntrinsicChain(const Value *LastUpdate, const Value *TargetPointer) {
+  IntrinsicChainInfo ChainInfo;
+  // We count backwards.
+  unsigned LastCalls = 0;
+  // Let's expose the real pointer.
+  TargetPointer = skipCast(TargetPointer);
+  if (!LastUpdate->getType()->isPointerTy())
+    return std::nullopt;
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+    if (auto *GEP = dyn_cast<GEPOperator>(LastUpdate)) {
+      if (GEP->hasAllZeroIndices()) {
+        // The result pointer and the first operand have the same value
+        LastUpdate = GEP->getPointerOperand();
+      } else {
+        return std::nullopt; // No pointer arithmetic is allowed.
+      }
+    } else if (Operator::getOpcode(LastUpdate) == Instruction::BitCast ||
+               Operator::getOpcode(LastUpdate) == Instruction::AddrSpaceCast) {
+      LastUpdate = cast<Operator>(LastUpdate)->getOperand(0);
+      if (!LastUpdate->getType()->isPointerTy())
+        return std::nullopt; // not a pointer.
+    } else {
+      if (auto *PHI = dyn_cast<PHINode>(LastUpdate)) {
+        // Reached final destination.
+        ChainInfo.IntrinsicCalls = ChainInfo.TotalIntrinsicCalls - LastCalls;
+        return ChainInfo;
+      }
+      if (auto *AIEObject = lookThroughAIEAddressingModes(LastUpdate)) {
+        IntrinsicArgsInfo NewArgSet = collectIntrinsicOperands(LastUpdate);
+        // Initialize counters.
+        if (!ChainInfo.IntrinsicParameters.CounterArgSet.empty() &&
+            !isDirectReachable(ChainInfo.IntrinsicParameters.CounterArgSet,
+                               AIEObject)) {
+          // if we already have counters, we need to reach the same
+          // AIEObject from the counters. Otherwise we can have some undesired
+          // counter update in between.
+          return std::nullopt; // not reachable.
+        }
+
+        // Update counters for the next round.
+        ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
+
+        // Initialize or compare other operands.
+        if (ChainInfo.IntrinsicParameters.ArgSet.empty()) {
+          ChainInfo.IntrinsicParameters.ArgSet = NewArgSet.ArgSet;
+        } else if (ChainInfo.IntrinsicParameters.ArgSet != NewArgSet.ArgSet) {
```

Comment:
Super-nit:
```
        if (!ChainInfo.IntrinsicParameters.ArgSet.empty() &&
            ChainInfo.IntrinsicParameters.ArgSet != NewArgSet.ArgSet) {
          // Constant iterator information mismatch between sequence of
          // intrinsic calls. We stop here.
          return std::nullopt;
        }

        // Update counters for the next round.
        ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
        ChainInfo.IntrinsicParameters.ArgSet = NewArgSet.ArgSet;
```

To be consistent with above.

Reply:
Thank, I think now it is a bit more readable!